### DISCUSSION TITLE
Cortana Binary

### DISCUSSION THREAD

**Discussion by coggy9** [coggy9/HKHacking#6] [General]
Now that I've got a shell session inside the Invoke, we can take a look at what the `cortana` binary does, and think about how to replace it with something else.

[Here's](https://pastebin.com/CyizyVG6) the output of just running the binary for a few seconds, I've removed any possibly sensitive IDs/info.

**Comment by rbeesley** [comment]
This is great progress. I'll have to go back and see how you're extracting the image and updating it, but it's very favorable that we can modify and that we aren't blocked from modifying it because of a signature check. At the very least what we should probably do is make an image which is the same as the Cortana final release, enable shell access, and then block the update checks when connecting it to a PC. At least this way we can see what it is doing today.

Thoughts for future updates, I'm decently happy with the speaker update, but I would want it to open the microphone as a communications BT instead of just music audio out. Second, being able to run another off the shelf assistant replacing Cortana, such as Alexa. And then there's an opportunity to build an API so that the assistant could be hosted on another machine and connecting to the speaker remotely.

Anyway, I haven't looked at this for over a week, so I'm excited about the progress. We probably need to figure out the toolchain to build programs to work with the Invoke kernel. The rest of that firmware will probably prove valuable to solving that.

**Reply by coggy9** [reply]
It's a very rudimentary process for now. 

- Extract the image with binwalk
- Extract the squashfs with the main system as root in an attempt to keep permissions
- Make edits
- Replace/inject new squashfs into image where old one was
- Flash with l2nand
- Fail due to bad CRC
- Fix CRC with the error l2nand gives
- Flash again


I'll release a 83_IMAGE with shell access and updates blocked in a few days, once I figure out the best way to block updates that won't cause error spam in the logs I know of.
